src/uu/sort/src/sort.rs

```diff
@@ -454,72 +519,196 @@ fn compare_by(a: &str, b: &str, settings: &Settings) -> Ordering {
         (a, b)
     };
 
-    for compare_fn in &settings.compare_fns {
-        let cmp: Ordering = if settings.random {
-            random_shuffle(a, b, settings.salt.clone())
+    // 1st Compare 
+    let mut cmp: Ordering = if settings.random {
+        random_shuffle(a, b, settings.salt.clone())
+    } else {
+        (settings.compare_fn)(a, b)
+    };
+
+    // Call "last resort compare" on any equal
+    if cmp == Ordering::Equal {
+        if settings.random || settings.stable || settings.unique {
+            cmp = Ordering::Equal
         } else {
-            compare_fn(a, b)
+            cmp = default_compare(a, b)
         };
-        if cmp != Ordering::Equal {
-            if settings.reverse {
-                return cmp.reverse();
-            } else {
-                return cmp;
-            }
-        }
+    };
+
+    if settings.reverse {
+        return cmp.reverse()
+    } else {
+        return cmp
     }
-    Ordering::Equal
 }
 
+// Test output against BSDs and GNU with env var 
+// lc_ctype=utf-8 until locales are implemented
+#[inline(always)]
 fn default_compare(a: &str, b: &str) -> Ordering {
     a.cmp(b)
 }
 
-fn get_leading_number(a: &str) -> &str {
+#[inline(always)]
+fn leading_num_common(a: &str) -> &str {
     let mut s = "";
-    for c in a.chars() {
-        if !c.is_numeric() && !c.eq(&'-') && !c.eq(&' ') && !c.eq(&'.') && !c.eq(&',') {
-            s = a.trim().split(c).next().unwrap();
+    // Strip string
+    for c in a.to_uppercase().chars() {
+        if !c.is_numeric()
+            && !c.is_whitespace()
+            && !c.eq(&DECIMAL_PT)
+            && !c.eq(&THOUSANDS_SEP)
+            && !c.eq(&ENOTATION)
+            && !a.chars().nth(0).unwrap_or('\0').eq(&POSITIVE)
+            && !a.chars().nth(0).unwrap_or('\0').eq(&NEGATIVE)
+        {
+            s = a.split(c).next().unwrap_or("");
             break;
         }
-        s = a.trim();
+        s = a;
     }
-    return s;
+    s
```

Comment:
Could you explain what this function is doing? It is not immediately clear to me. Also, that `to_uppercase` seems unneeded and a major performance problem because it allocates a new string each time.

Reply:
It's splitting the string at its non-numeric boundary.  So 156890BNNN becomes 156890.

Re: performance, what can I say?  You're right.   See below.  to_uppercase was only there for the ENOTATION test (is the next char either e or E). 

Benchmark #1: cat ~/Programming/countwords/kjvbible.txt  | ./target/release/sort -n
  Time (mean ± σ):     231.7 ms ±   3.8 ms    [User: 821.8 ms, System: 43.7 ms]
  Range (min … max):   227.9 ms … 241.4 ms    12 runs
 
Benchmark #2: cat ~/Programming/countwords/kjvbible.txt  | gsort -n
  Time (mean ± σ):      1.224 s ±  0.060 s    [User: 1.187 s, System: 0.031 s]
  Range (min … max):    1.132 s …  1.291 s    10 runs
 
Summary
  'cat ~/Programming/countwords/kjvbible.txt  | ./target/release/sort -n' ran
    5.28 ± 0.27 times faster than 'cat ~/Programming/countwords/kjvbible.txt  | gsort -n'

Reply:
I've modified but I'm having an issue with test_exponents_positive.  I'll push when I've figured it out.

Reply:
Oh, that makes sense. You could use `char_indices` to make this a bit clearer.
i.e.
```rust
for (idx, c) in a.char_indices() {
        if !c.is_numeric()
            && !c.is_whitespace()
            && !c.eq(&DECIMAL_PT)
            && !c.eq(&THOUSANDS_SEP)
            && !c.eq(&ENOTATION)
            && !a.chars().nth(0).unwrap_or('\0').eq(&POSITIVE)
            && !a.chars().nth(0).unwrap_or('\0').eq(&NEGATIVE)
        {
            s = a[..idx];
```

Reply:
Done and done.  Thanks.